gateway/ln-gateway/src/gateway_module_ng/send_sm.rs

```diff
@@ -0,0 +1,183 @@
+use fedimint_core::task::sleep;
+use fedimint_core::{Amount, OutPoint};
+use fedimint_ln_client_ng::LightningClientStateMachines;
+use fedimint_ln_common_ng::contracts::OutgoingContract;
+use fedimint_ln_common_ng::{LightningInput, OutgoingWitness, Witness};
+use lightning_invoice::Bolt11Invoice;
+use tracing::error;
+
+use crate::gateway_lnrpc::PayInvoiceRequest;
+use crate::gateway_module_ng::GatewayClientContextNG;
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Decodable, Encodable)]
+pub struct SendStateMachine {
+    pub operation_id: OperationId,
+    pub contract: OutgoingContract,
+    pub state: SendSMState,
+}
+
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Decodable, Encodable)]
+pub enum SendSMState {
+    Sending(Sending),
+    Claiming([u8; 32], Vec<OutPoint>),
+    Cancelled,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Decodable, Encodable)]
+pub struct Sending {
+    pub max_delay: u64,
+    pub max_fee_msat: u64,
+    pub invoice: Bolt11Invoice,
+    pub claim_keypair: KeyPair,
+}
+
+impl State for SendStateMachine {
+    type ModuleContext = GatewayClientContextNG;
+
+    fn transitions(
+        &self,
+        context: &Self::ModuleContext,
+        global_context: &DynGlobalClientContext,
+    ) -> Vec<StateTransition<Self>> {
+        let gc = global_context.clone();
+
+        match &self.state {
+            SendSMState::Sending(sending) => {
+                vec![StateTransition::new(
+                    Self::send_payment(
+                        context.clone(),
+                        sending.max_delay,
+                        sending.max_fee_msat,
+                        sending.invoice.clone(),
+                    ),
+                    move |dbtx, result, old_state| {
+                        Box::pin(Self::transition_send_payment(
+                            dbtx,
+                            old_state,
+                            gc.clone(),
+                            result,
+                        ))
+                    },
+                )]
+            }
+            SendSMState::Claiming(..) => {
+                vec![]
+            }
+            SendSMState::Cancelled => {
+                vec![]
+            }
+        }
+    }
+
+    fn operation_id(&self) -> OperationId {
+        self.operation_id
+    }
+}
+
+impl SendStateMachine {
+    async fn send_payment(
+        context: GatewayClientContextNG,
+        max_delay: u64,
+        max_fee_msat: u64,
+        invoice: Bolt11Invoice,
+    ) -> Result<[u8; 32], ()> {
+        let lightning_context = loop {
+            match context.gateway.get_lightning_context().await {
+                Ok(lightning_context) => break lightning_context,
+                Err(_) => {
+                    sleep(std::time::Duration::from_secs(1)).await;
+                }
+            }
+        };
+
+        if lightning_context.lightning_public_key == invoice.recover_payee_pub_key() {
```

Comment:
Ok so this is checking if the gateway's lightning node is the same as the lightning node that generated the invoice, then we can do an internal swap and just start the receive state machine. A comment here would help.

Will this still work if there are > 1 gateways connected to the same lightning node? (we don't have explicit support for this today, but we might want to in the future for custodial payment providers for instance).